File: ani/arrays/2_sum.py
def two_sum(nums, target):
    mp = {}  # value -> index
    
    for i, num in enumerate(nums):
        diff = target - num
        
        if diff in mp:
            return [mp[diff], i]  # Found the pair
        
        mp[num] = i  # Store index of current number

    return []  # If no pair found

# Example:
print(two_sum([2,7,11,15], 9))  # Output: [0,1]

File: ani/arrays/2_sum.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
using namespace std;

vector<int> twoSum(vector<int>& nums, int target) {
    unordered_map<int, int> mp; // value -> index

    for(int i = 0; i < nums.size(); i++) {
        int diff = target - nums[i];

        // If complement exists → found the pair
        if(mp.find(diff) != mp.end()) {
            return {mp[diff], i};
        }

        mp[nums[i]] = i; // Store index
    }

    return {}; // No solution case
}

int main() {
    vector<int> nums = {2,7,11,15};
    int target = 9;

    vector<int> result = twoSum(nums, target);
    cout << result[0] << " " << result[1]; // Output: 0 1
}
